addons/web_editor/static/src/js/wysiwyg/widgets/generation_dialog.js

```diff
@@ -0,0 +1,177 @@
+/** @odoo-module **/
+
+import { Component, useState } from "@odoo/owl";
+import { useAutofocus } from "@web/core/utils/hooks";
+import { Dialog } from "@web/core/dialog/dialog";
+
+export class GenerationDialog extends Component{
+    static components = { Dialog };
+    static props = {
+        mode: { validate: e => ['prompt', 'alternatives'].includes(e) },
+        initialPrompt: { type: String, optional: true },
+        context: { type: String, optional: true },
+        rpcService: Function,
+        restore: Function,
+        close: Function,
+        confirm: { type: Function, optional: true },
+        insert: Function,
+        onError: Function,
+        onFinally: { type: Function, optional: true },
+    };
+    static defaultProps = {
+        initialPrompt: '',
+        context: '',
+        confirm: () => {},
+        onFinally: () => {},
+    }
+    static template = 'web_edior.GenerationDialog';
+
+    async setup() {
+        this.isConfirmedOrCancelled = false; // ensures we do not confirm and/or cancel twice
+        this.state = useState({
+            prompt: this.props.initialPrompt,
+            generations: Array(5).fill().map(() => ({})),
+            selectedGeneration: null,
+        });
+        useAutofocus({ refName: 'promptInput' });
+        if (this.props.mode === 'alternatives') {
+            const conversationHistory = [
+                {
+                    role: 'system',
+                    content: 'You are a helpful assistant, your goal is to help the user write alternatives to their text.',
+                },
+                {
+                    role: 'user',
+                    content:  'Consider the following text : \n' + this.state.prompt,
+                },
+                {
+                    role: 'assistant',
+                    content: 'Thanks for this text, what do you need me to do with it?',
+                },
+                {
+                    role: 'user',
+                    content:  'I need you to write one alternative version of my text.',
+                },
+            ];
+            for (const generation of this.state.generations) {
+                const content = await this._generate(this.state.prompt, conversationHistory, content => {
+                    generation.text = content;
+                    conversationHistory.push({
+                        role: 'assistant',
+                        content,
+                    }, {
+                        role: 'user',
+                        content: 'Thanks. I need another one.',
+                    });
+                }).catch(() => {
+                    this.state.generations = [];
+                });
+                if (!content) {
+                    break;
+                }
+            }
+        }
+    }
+
+    selectGeneration(ev) {
+        this.state.selectedGeneration = +ev.target.getAttribute('data-generation-index');
+    }
+
+    parseGeneratedContent(content) {
```

Comment:
Now I split the class in three, it's in the common one and so the others can access it with `this`. Otherwise I could still take it out and export it separately. What would you prefer?

Reply:
The question we need to ask ourselves is if it might be useful to someone to override or patch this function. In a way I think it is, if they want to support other stuff coming from chatgpt...

Reply:
Yes that sounds good.